library/std/src/io/buffered/switchwriter.rs

```diff
@@ -0,0 +1,115 @@
+use crate::fmt::Arguments;
+use crate::io::{self, buffered::LineWriterShim, BufWriter, IoSlice, Write};
+/// Different buffering modes a writer can use
+#[unstable(feature = "stdout_switchable_buffering", issue = "78515")]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum BufferMode {
+    /// Unbuffered: forward writes directly to the underlying writer. In some
+    /// cases, a writer may buffer temporarily (for instance, when processing
+    /// a formatted write), but even in this case it will always immediately
+    /// flush.
+    None,
+
+    /// Block buffering: buffer writes until the buffer is full, then forward
+    /// to the underlying writer
+    Block,
```

Comment:
Would adding the block size here be useful? (e.g. `BufferMode::Block { size: usize }`) Doing this as an enum means things like that can't be added once it is stable.

Reply:
I like this idea in principle; the trouble is that it complicates the interface of `set_mode`, because we need to do something if you try to shrink the buffer while there's still stuff in it.

Reply:
stdio.h's setvbuf is fallible (though it is not clear from the standard when a non-zero value actually comes back, it only says "if an invalid value is given for `mode` or the request cannot be honored"), and it is UB to invoke it after a write.
Hmm...

> If stream is NULL, or if mode or size is not within a valid change, the invalid parameter handler is invoked, as described in Parameter Validation. If execution is allowed to continue, this function returns -1 and sets errno to EINVAL.

Ah, I see. Yes, "if the stream you are trying to adjust the buffering to does not exist anymore" is one good reason.

Reply:
I'm leaning towards keeping them separate, since mode changes (as currently designed) are trivial and doesn't touch the buffer, but modifying the buffer size is not.

Reply:
Why would line buffering take a value?

Reply:
> Setting the mode to Immediate under this scheme might give the impression that the buffer storage is flushed & freed, which I want to avoid.

How about `Block(0)`? Would that flush and free the storage?

Why do you want to avoid this? Stdout wouldn't use the buffer for `write_fmt` anyway, so the buffer is just sitting there, doing nothing.

If `Immediate` also uses a buffer (for `write_fmt`), shouldn't it also have a size?

If you're supporting different buffer sizes, I don't think it makes much sense to still keep this `write_fmt` optimization around, as it mostly complicates the design and the code with no clear benefit, since stdout isn't using it in practice anyway.

Reply:
> Why would line buffering take a value?

It'd be the size of the buffer; line buffering will still flush incomplete lines if they exceed the size of the buffer.

> How about `Block(0)`? Would that flush and free the storage?

This is exactly why I'm hesitant to make this change; my inclination is that switching the buffer mode should not be fallible.

> as it mostly complicates the design and the code with no clear benefit.

It's purely in there based on `SwitchWriter`, which was designed independent of its usage in `stdout`. I'd be okay to remove it on this basis since right now `SwitchWriter` isn't exported publicly, but that was the original rationale.

> If you're supporting different buffer sizes

Right now I'm not; the precedent I'm following is that the buffer size is fixed after the adapter is constructed.

Reply:
To review:
- Adjusting the buffer size should clearly be fallible
   - We *could* add additional semantics to setting the buffer size so that it is infallible
   - However, that is essentially preselecting an error handling strategy. This overcomplicates the API.
- There is clearly a need to be able to set a different buffering strategy.
- There is most likely a need to be able to adjust the buffer size (especially when switching strategies)
- So there is clearly a need to be able to set both strategy and size in a single method call.

I believe that we would agree on this as our set of constraints?

Reply:
> So there is most likely a need to be able to set both strategy and size in a single method call.

I don't see how this follows. If you need both is there a benefit of having them in one method instead of calling both consecutively and perhaps also flushing before doing that?

Reply:
That presumes that there is no difference in number of syscalls generated, which I do not believe is the case.

Reply:
Once the buffer is flushed (either by explicit flush or shrinking it) additional mode changes shouldn't incur any syscalls, no?

Reply:
> That presumes that there is no difference in number of syscalls generated, which I do not believe is the case.

I'm presuming this on the basis that only changing the *mode* doesn't generate any syscalls at all (under the current design). I can see an argument that this shouldn't be the case- for instance, it could be argued that switching to unbuffered mode should flush immediately- but I found I preferred a design where as much as possible the i/o is constrained to explicit i/o method calls.

In fact, having now said it like that, I think I'd even like a model where shrinking the buffer doesn't attempt to flush, but instead simply fails if there is too much buffered data, and the use needs to separately flush on their own.

Reply:
On the other hand, there's existing precedence for "flush when buffer is shrunk" in the form of `into_inner`, which attempts to flush the buffer before dropping it and returning the wrapped writer.